kernel/src/vm/vmar/mod.rs

```diff
@@ -277,6 +326,65 @@ impl VmarInner {
 
         return_errno_with_message!(Errno::ENOMEM, "Cannot find free region for mapping");
     }
+
+    /// Splits and unmaps the found mapping if the new size is smaller.
+    /// Enlarges the last mapping if the new size is larger.
+    fn resize_mapping(
+        &mut self,
+        vm_space: &VmSpace,
+        map_addr: Vaddr,
+        old_size: usize,
+        new_size: usize,
+        rss_delta: &mut RssDelta,
+    ) -> Result<()> {
+        debug_assert!(map_addr % PAGE_SIZE == 0);
+        debug_assert!(old_size % PAGE_SIZE == 0);
+        debug_assert!(new_size % PAGE_SIZE == 0);
+
+        // FIXME: We should check whether all existing ranges in
+        // `map_addr..map_addr + old_size` have a mapping. If not,
+        // we should return a `Err`.
+
+        if new_size == 0 {
+            return_errno_with_message!(Errno::EINVAL, "can not resize a mapping to 0 size");
+        }
+
+        if new_size == old_size {
+            return Ok(());
+        }
+
+        let old_map_end = map_addr + old_size;
+        let new_map_end = map_addr + new_size;
+
+        if new_size < old_size {
+            self.alloc_free_region_exact_truncate(
+                vm_space,
+                new_map_end,
+                old_map_end - new_map_end,
+                rss_delta,
+            )?;
+            return Ok(());
+        }
+
+        let last_mapping = self.vm_mappings.find_one(&(old_map_end - 1)).unwrap();
+        let last_mapping_addr = last_mapping.map_to_addr();
+        let extra_mapping_start = last_mapping.map_end();
+
+        self.check_extra_size_fits_rlimit(new_map_end - extra_mapping_start)?;
+
+        let last_mapping = self.remove(&last_mapping_addr).unwrap();
+
+        if let Err(e) =
+            self.alloc_free_region_exact(extra_mapping_start, new_map_end - extra_mapping_start)
```

Comment:
Following the above comments, to fail with errors for the case where `old_map_end < extra_mapping_start`, I think `alloc_free_region_exact` should
 - be changed to `alloc_free_region_exact(old_map_end, new_map_end - old_map_end)`, and
 - happen before `self.remove(&last_mapping_addr)`.